shm/ship.cpp

```diff
@@ -0,0 +1,179 @@
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+#include "ship.hpp"
+#include "player.hpp"
+#include "fruit.hpp"
+#include "alcohol.hpp"
+#include "item.hpp"
+
+Ship::Ship(Time *time)
+    : id_(-1), time_(time)
+{
+    time_->attach(this);
+}
+
+Ship::Ship(int capacity, int Crew, int speed, const std::string &name, size_t id, Time *time)
+    : capacity_(capacity), crew_(Crew), speed_(speed), name_(name), id_(id), time_(time)
+{
+    time_->attach(this);
+    shipCargo.reserve(10);
+}
+
+Ship::Ship(int maxCrew, int speed, size_t id, Time *time)
+    : Ship(0, maxCrew, speed, "", id, time)
+{
+    time_->attach(this);
+}
+
+Ship::~Ship()
+{
+    time_->detach(this);
+}
+
+Ship &Ship::operator-=(size_t num)
+{
+    if (crew_ < num)
+    {
+        crew_ = 0;
+    }
+    else
+    {
+        crew_ -= num;
+    }
+    return *this;
+}
+
+Ship &Ship::operator+=(size_t num)
+{
+    crew_ += num;
+    return *this;
+}
+
+void Ship::addCargo(Cargo * item)
+{
+    auto cargoPtr = findMatchCargo(item);
+    if(findMatchCargo(item))
+    {
+        findMatchCargo(item)->increaseAmount(item->getAmount());
+    }
+    else
+    {
+        shipCargo.emplace_back(item);
+    }
+}
+
+void Ship::removeCargo(Cargo * item, size_t amount)
+{
+    auto shipCargoAmount = findMatchCargo(item)->getAmount();
+    if(shipCargoAmount == amount)
+    {
+        auto it = std::find_if(begin(shipCargo), end(shipCargo), [item](const auto* el){ return *el == *item; });
+        if(it != end(shipCargo)) {
+            shipCargo.erase(it);
+        }
+    }
+    else
+    {
+        findMatchCargo(item)->reduceAmount(amount);
+    }
+}
+
+Cargo * Ship::findMatchCargo(Cargo * cargo)
+{
+    for (auto &el : shipCargo)
+    {
+        if (*el == *cargo)
+        {
+            return el;
+        }
+    }
+    return nullptr;
+}
+void Ship::load(Cargo * loadCargo, size_t amount)
+{
+    if (amount <= calculateAvailableSpace())
+    {        
+        loadCargo->reduceAmount(amount);
+        Cargo* toAdd;
+        if(Fruit* f = dynamic_cast<Fruit*>(loadCargo)){
+            toAdd = new Fruit(*f);
+
+        }
+        if(Item* i = dynamic_cast<Item*>(loadCargo)){
+            toAdd = new Item(*i);
+        }
+        if(Alcohol* a = dynamic_cast<Alcohol*>(loadCargo)){
+            toAdd = new Alcohol(*a);
+        }        
```

Comment:
new is deprecated. Please create 'clone()` method to cargo, which will clone itself and return unique_ptr to the new object. When you do this, you avoid ugly dynamic_cast which in 95% show the wrong style of code.

`const auto* toAdd = loadCargo->clone()`